mlrun/model.py

```diff
@@ -56,22 +60,115 @@ def _verify_dict(param, name, new_type=None):
             return new_type.from_dict(param)
         return param
 
-    def to_dict(self, fields=None, exclude=None):
-        """convert the object to a python dictionary"""
+    def to_dict(
+        self, fields: list = None, exclude: list = None, strip: bool = False
+    ) -> dict:
+        """
+        convert the object to a dict
+        :param fields:  A list of fields to include in the dictionary. If not provided, the default value is taken
+        from `self._dict_fields` or from the object __init__ params.
+        First we run over the fields and serialize them, then we run over the fields and enrich them.
+        :param exclude: A list of fields to exclude from the dictionary. If not provided, the default value is an
+         empty list and the object's `_fields_to_exclude_for_serialization` attribute is appended to it.
+        :param strip:  If True, the object's `_fields_to_strip` attribute is appended to the exclude list.
+        :return: A dictionary representation of the object.
+        """
         struct = {}
-        fields = fields or self._dict_fields
-        if not fields:
-            fields = list(inspect.signature(self.__init__).parameters.keys())
-        for t in fields:
-            if not exclude or t not in exclude:
-                val = getattr(self, t, None)
-                if val is not None and not (isinstance(val, dict) and not val):
-                    if hasattr(val, "to_dict"):
-                        val = val.to_dict()
-                        if val:
-                            struct[t] = val
-                    else:
-                        struct[t] = val
+
+        fields = self._resolve_fields(fields)
+        fields_to_exclude = exclude or []
+        if strip:
+            fields_to_exclude += self._fields_to_strip
+
+        fields_to_save = (
+            set(fields)
+            - set(fields_to_exclude)
+            - set(self._fields_to_exclude_for_serialization)
+            - set(self._fields_to_exclude_for_enrichment)
+        )
+
+        for field in fields_to_save:
+            field_value = getattr(self, field, None)
+            if self._is_valid_field_value(field_value):
+                if hasattr(field_value, "to_dict"):
+                    field_value = field_value.to_dict(strip=strip)
+                    if self._is_valid_field_value(field_value):
+                        struct[field] = field_value
+                else:
+                    struct[field] = field_value
+
+        # if field is in field_to_exclude_for_serialization is also in the fields_to_exclude list,
+        # then no need to iterate over it
+        fields_to_serialize = list(
+            set(self._fields_to_exclude_for_serialization) - set(fields_to_exclude)
```

Comment:
So, I'm a bit confused. Is the `_fields_to_exclude_for_serialization` a list of fields to exclude or fields to strip? It's not a good name, I fear.

Reply:
renamed and added documentation